src/middlewared/middlewared/plugins/datastore.py

```diff
@@ -145,245 +338,227 @@ def query(self, name, filters=None, options=None):
               "params": ["account.bsdusers", [ ["username", "=", "root" ] ], {"get": true}]
             }
         """
-        model = self.__get_model(name)
-        if options is None:
-            options = {}
-        else:
-            # We do not want to make changes to original options
-            # which might happen with "prefix"
-            options = options.copy()
+        table = self._get_table(name)
+
+        # We do not want to make changes to original options
+        # which might happen with "prefix"
+        options = options.copy()
 
-        qs = model.objects.all()
+        aliases = {}
+        if options['count']:
+            qs = select([func.count(self._get_pk(table))])
+        else:
+            columns = list(table.c)
+            from_ = table
+            if options['relationships']:
+                aliases = self._get_queryset_joins(table)
+                for foreign_key, alias in aliases.items():
+                    columns.extend(list(alias.c))
+                    from_ = from_.outerjoin(alias, alias.c[foreign_key.column.name] == foreign_key.parent)
 
-        extra = options.get('extra')
-        if extra:
-            qs = qs.extra(**extra)
+            qs = select(columns).select_from(from_)
 
-        prefix = options.get('prefix')
+        prefix = options['prefix']
```

Comment:
If `options` is `null` and we recreate the dict, it wouldn't have `prefix`. Same is true for below usages